File: src/main/java/dev/liquid/osamaclient/feature/impl/esp/ParticleESP.java
package dev.liquid.osamaclient.feature.impl.esp;

import dev.liquid.osamaclient.config.OsamaClientConfig;
import dev.liquid.osamaclient.event.ParticleSpawnEvent;
import dev.liquid.osamaclient.feature.IFeature;
import dev.liquid.osamaclient.util.RenderUtil;
import lombok.Getter;
import net.minecraft.util.EnumParticleTypes;
import net.minecraft.util.Vec3;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import java.awt.*;
import java.util.Iterator;
import java.util.LinkedList;

public class ParticleESP implements IFeature {
  public static ParticleESP instance;
  private final LinkedList<Particle> particleRenderList = new LinkedList<>();

  public static ParticleESP getInstance() {
    if (instance == null) {
      instance = new ParticleESP();
    }
    return instance;
  }

  @Override
  public String getFeatureName() {
    return "Particle Esp";
  }

  @Override
  public boolean canEnable() {
    return OsamaClientConfig.particleEsp && this.isPassiveFeature();
  }

  @Override
  public void disable() {
    if (!this.canEnable()) return;
    OsamaClientConfig.particleEsp = false;
  }

  @Override
  public boolean isPassiveFeature() {
    return true;
  }

  @SubscribeEvent
  public synchronized void onParticleSpawn(ParticleSpawnEvent event) {
    if (!this.canEnable()) return;

    EnumParticleTypes targetParticleType = EnumParticleTypes.getParticleFromId(OsamaClientConfig.particleEspParticleID);
    if (event.getParticleType() != targetParticleType) return;
    this.particleRenderList.push(new Particle(event.getParticlePosition(), OsamaClientConfig.particleEspParticleTime));
  }

  @SubscribeEvent
  public synchronized void onRenderWorldLast(RenderWorldLastEvent event) {
    if (!this.canEnable()) return;
    if (this.particleRenderList.isEmpty()) return;

    Iterator<Particle> iterator = particleRenderList.iterator();
    while (iterator.hasNext()) {
      Particle particle = iterator.next();
      RenderUtil.drawPoint(particle.getPos(),
          OsamaClientConfig.particleEspEspColor.toJavaColor(),
          OsamaClientConfig.particleEspParticleSize);
      particle.duration--;
      if (particle.duration <= 0) {
        iterator.remove();
      }
    }
  }
}

@Getter
class Particle {
  private final Vec3 pos;
  int duration;

  public Particle(Vec3 pos, int duration) {
    this.pos = pos;
    this.duration = duration;
  }
}

File: src/main/kotlin/dev/liquid/osamaclient/feature/implementation/esp/ParticleESP.kt
package dev.liquid.osamaclient.feature.implementation.esp

import dev.liquid.osamaclient.event.PacketEvent
import dev.liquid.osamaclient.event.ParticleSpawnEvent
import dev.liquid.osamaclient.feature.AbstractFeature
import dev.liquid.osamaclient.util.RenderUtil.drawPoint
import dev.liquid.osamaclient.util.config
import net.minecraft.util.EnumParticleTypes
import net.minecraft.util.Vec3
import net.minecraftforge.client.event.ClientChatReceivedEvent
import net.minecraftforge.client.event.RenderWorldLastEvent
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent
import java.util.*

class ParticleESP : AbstractFeature() {
  // Singleton Creator
  companion object {
    private var instance: ParticleESP? = null
    fun getInstance(): ParticleESP {
      if (instance == null) {
        instance = ParticleESP()
      }
      return instance!!
    }
  }

  override val featureName = "Particle ESP"
  override val isPassiveFeature = true
  private val particleRenderList = LinkedList<Particle>()

  override fun canEnable() = config.particleEsp && isPassiveFeature
  override fun disable() {
    if (!this.canEnable()) return
    config.particleEsp = false
  }


  @SubscribeEvent
  @Synchronized
  fun onParticleSpawn(event: ParticleSpawnEvent) {
    if (!this.canEnable()) return
    val targetParticleType = EnumParticleTypes.getParticleFromId(config.particleEspParticleID)
    if (event.particleType !== targetParticleType) return
    this.particleRenderList.push(Particle(event.particlePosition, config.particleEspParticleTime))
  }

  @SubscribeEvent
  @Synchronized
  override fun onRenderWorldLastEvent(event: RenderWorldLastEvent) {
    if (!this.canEnable()) return
    if (this.particleRenderList.isEmpty()) return
    val iterator = this.particleRenderList.iterator()
    while (iterator.hasNext()) {
      val particle = iterator.next()
      drawPoint(
        particle.pos,
        config.particleEspEspColor.toJavaColor(),
        config.particleEspParticleSize
      )
      particle.duration--
      if (particle.duration <= 0) {
        iterator.remove()
      }
    }
  }

  override fun onTick(event: TickEvent.ClientTickEvent) {
  }
  override fun onPacketReceive(event: PacketEvent.Received) {
  }
  override fun onChatReceive(event: ClientChatReceivedEvent) {
  }
}

data class Particle(val pos: Vec3, var duration: Int)